app/services/stats/base_monthly_performance_report.rb

```diff
@@ -37,13 +39,45 @@ def self.description
       raise '#description should be defined in sub-class of BaseMonthlyPerformanceReport'
     end
 
-    def run(*)
+    def process(offset, report_job_guid=nil)
       CaseSelector.new(case_scope)
-        .cases_received_in_period(@period_start, @period_end)
-        .includes(:responded_transitions, :approver_assignments, :assign_responder_transitions)
-        .each { |kase| analyse_case(kase) }
+      .cases_received_in_period(@period_start, @period_end)
+      .order(:id)
+      .limit(ROWS_PER_FRAGMENT)
+      .offset(offset)
+      .includes(:responded_transitions, :approver_assignments, :assign_responder_transitions) 
+      .each { |kase| analyse_case(kase) }
+      if !report_job_guid.nil?
+        redis = Redis.new
+        redis.set(report_job_guid, @stats.stats.to_json)
+      end
 
-      @stats.finalise
+    end
+
+    def run(*)
+      offset = 0 
```

Comment:
This is quite a long method - might be worth having some private methods with some descriptive names that make it clear the steps that are happening here - something like:

```ruby
if data_size > ROWS_PER_FRAGMENT
    batch_up_records()
end
```
```